Enabled ASAN by default while running sanitycheck for native_posix

vanwinkeljan opened the pull request:
Enable address sanitizer for CI test runs

Fixes: #19012 

~Blocked by #20160~

aescolar reviewed (dismissed):
Would seem you missed this one :) :
https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/sanitycheck#L1339

vanwinkeljan commented:
> Would seem you missed this one :)

Yes noticed the same yesterday :)

SebastianBoe commented on boards/posix/native_posix/Kconfig.defconfig:
```diff
@@ -103,3 +103,14 @@ config USB_NATIVE_POSIX
+
+# i386 assembly code used in MBEDTLS does not compile with speed optimization
+# if address sanitizer is enabled, as such switch default optimization level
+# to speed
```
This comment doesn't make sense.

It says "We can't use X when we do Y, therefore do Y."

vanwinkeljan replied:
Oops comment should read `i386 assembly code used in MBEDTLS does not compile with size optimization ...`

aescolar commented on boards/posix/native_posix/Kconfig.defconfig:
```diff
@@ -103,3 +103,14 @@ config USB_NATIVE_POSIX
+if ASAN && !64BIT && MBEDTLS
+
+choice OPTIMIZATION_LEVEL
+	default SPEED_OPTIMIZATIONS
```
It would then be the same issue with the nrf52_bsim (even if we typically do not test mbedtls there)
Maybe it would be easier to just set the default optimization level to no_opt with ASAN just like with COVERAGE?

SebastianBoe replied:
I didn't want to nitpick, but if we are going to change this then it would be cleaner to just modify the build environment for the mbedtls library, and not for the entire build.

vanwinkeljan replied:
An option would be to apply the following patch to `crypto/mbedtls/CMakeLists.txt`

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index d382aea..ccf9a61 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -94,6 +94,10 @@ if(CONFIG_MBEDTLS_BUILTIN)
   zephyr_library_sources(library/x509write_csr.c)
   zephyr_library_sources(library/xtea.c)

+if(CONFIG_ARCH_POSIX AND CONFIG_ASAN AND NOT CONFIG_64BIT)
+  set_property(SOURCE library/bignum.c APPEND PROPERTY COMPILE_OPTIONS "-O0")
+endif ()
+
   zephyr_library_link_libraries(mbedTLS)
 else()
   assert(CONFIG_MBEDTLS_LIBRARY "MBEDTLS was enabled, but neither BUILTIN or LIBRARY was selected.")

```

SebastianBoe replied:
Change from -O0 to ${OPTIMIZE_FOR_SPEED_FLAG} or ${OPTIMIZE_FOR_NO_OPTIMIZATIONS_FLAG} and this looks good to me.

vanwinkeljan replied:
changes moved to mbedtls module

vanwinkeljan commented:
blocked by mbedtls module PR https://github.com/zephyrproject-rtos/mbedtls/pull/4

vanwinkeljan commented:
Updated ASAN_OPTIONS to disable leak detection, if leak detection is enabled deadlocks occur on exit

vanwinkeljan commented:
Bumped mbedtls module reference to take in:
https://github.com/zephyrproject-rtos/mbedtls/pull/4

andrewboie commented on scripts/sanitycheck:
```diff
@@ -1661,9 +1667,13 @@ class TestInstance:
 
+            if not options.disable_asan:
+                if platform.type == "native":
+                    content = content + "\nCONFIG_ASAN=y"
```
could we select this in Kconfig instead?

historically, having sanitycheck build differently than local builds of the same code has led to all kinds of confusion.

vanwinkeljan replied:
Yes this would be possible, but it would mean that ASAN is enabled by default for posix arch boards and require libasan to be installed.

Further I did observe deadlocks* while running sanitycheck with ASAN memory leak detection enabled, which is by default enabled, and it needs to be disabled by setting an environment variable. Enabling ASAN by default would mean that all user of the posix arch should be aware of this deadlock possibility and that they need to set an environment variable.

* Until now is I was not able to pinpoint the root cause of the deadlock but it points to a race condition between LSAN (memory leak detector in ASAN) running its memory leak detection and the the shutdown of the posix arch spawned threads.

andrewboie replied:
> Yes this would be possible, but it would mean that ASAN is enabled by default for posix arch boards and require libasan to be installed

Right, but since we're doing this in sanitycheck I think we have a libasan dependency already.

> (1) Until now is I was not able to pinpoint the root cause of the deadlock but it points to a race condition between LSAN (memory leak detector in ASAN) running its memory leak detection and the the shutdown of the posix arch spawned threads.

If there isn't a GH issue for this already please file one.

@aescolar any thoughts on this?

vanwinkeljan replied:
@nashif @aescolar Any input on enabling ASAN by default in Kconfig instead of enabling it only by default in sanitycheck?

@andrewboie 
> does native posix work in Windows?

No it does not work in windows, I dont know even how well native_posix is supported on mac

aescolar replied:
> Any input on enabling ASAN by default 

Sorry but that is a no go. This discussion already happened 1.5 years ago. Just like we do not run by default posix arch boards in valgrind. Compiling with ASAN is nice for native_posix in many cases, but is a no go in others. Even if everything else worked fine, it adds a lot of runtime overhead, which for long BLE simulations is a dealbreaker.

I certainly understand @andrewboie 's point, and I think it is valid. But then the solution should be to have this sanitycheck option inverted (to `--enable-asan`). Then have `run_ci.sh ` call it so it is utilized in CI, but not by default in local runs (if @andrewboie thinks it would be sensible, otherwise just have it as an option for people to run locally just as the valgrind one)

> does native posix work in Windows?

Nope, Windows in a not a POSIX compliant OS :)

> I dont know even how well native_posix is supported on mac

It is known to not work (see 10945)

vanwinkeljan replied:
@andrewboie Issue created for deadlock: #20122

aescolar reviewed (changes requested):
Sorry for coming with this so late feedback, but I do agree with @andrewboie, I think this option should be inverted to `--enable-asan` and the default should be to run with it off.

vanwinkeljan commented:
> Sorry for coming with this so late feedback, but I do agree with @andrewboie, I think this option should be inverted to --enable-asan and the default should be to run with it off.

Ok I will change it to `--enable-asan` and update the run_ci.sh script to enable it during sanitycheck runs.

nashif reviewed (changes requested):
on a second though and realizing it is enabled by default I agree with others here. I also change the valgrind PR to be enabled using --enable-valgrind for consistency.

aescolar commented:
@nashif @vanwinkeljan By the way, these 2 options are mutually exclusive, so whoever merges after the other, please add a note in the help and check that prevents running with both on.

vanwinkeljan commented:
@nashif @aescolar @andrewboie Before I update the run_ci.sh script, I assume we still want to enable ASAN for CI test runs?

vanwinkeljan commented:
Updated PR:
1) Changed `--disable-asan` to `--enable-asan`
2) Added option to enable leak sanitizer (`--enable-lsan`)
3) Made  `--enable-asan` and `--enable-valgrind` mutual exclusive options
4) Update `run_ci.sh` to enable ASAN for CI test runs

andrewboie commented:
I'm not gonna stand in the way with a -1 but this still has the same problem: CI runs are going with a different configuration as a local build, or even a local sanitycheck run. I *hate* this.

@aescolar @nashif how about enabling ASAN by default if (and only if) CONFIG_TEST is enabled? Similar to how we turn some other non-default options on purely for test cases like minimal logging and HW stack protection.

aescolar commented:
Let's do one thing. Let's start by having this PR only add the options to sanitycheck but not changing any default behavior or enabling it in CI. That is fully uncontroversial, and will also enable some more easy mileage on this feature.
After we can take as a 2nd round the change of CI or defaults, when/if we are sure enough there is no issues left.

vanwinkeljan commented:
@aescolar This PR is about enabling ASAN by default in CI to detect issues if this is not done new issues will be introduced and another round of clean up can be done if we would like to enable it (see #19012 for the list of PRs fixing issues detected by ASAN).

aescolar commented on scripts/sanitycheck:
```diff
@@ -581,7 +581,14 @@ class BinaryHandler(Handler):
+        # Disable memory leak detection which leads to unexpected deadlocks
+        # in posix architecture
+        if not options.enable_lsan:
+            env["ASAN_OPTIONS"] += "detect_leaks=0"
```
Should this not be under a  `if options.enable_asan:`?
It is a bit nitpicky, but a user be enabling ASAN by other means (than with the --enable-asan option) and not expect sanitycheck to mingle with their environment behind their backs.

vanwinkeljan replied:
fixed in #20160

aescolar commented:
> @aescolar This PR is about enabling ASAN 

I understand that.
But @andrewboie 's comment is sensible enough like to hold this PR due to the change in the CI side. Yet it would be nice to get the sanitycheck change in, that part seems non controversial and nice to get in ASAP (would be nice to get that small fix I ask for though)
So, would you then prefer to create a separate PR with only the sanitycheck change, which we could merge already now?

vanwinkeljan commented:
> So, would you then prefer to create a separate PR with only the sanitycheck change, which we could merge already now?

That was my idea, I will do this over the weekend (including the change request)

vanwinkeljan commented:
@aescolar Moved the enable options in a separate PR (#20160)

andrewboie commented:
> But @andrewboie 's comment is sensible enough like to hold this PR due to the change in the CI side. Yet it would be nice to get the sanitycheck change in, that part seems non controversial and nice to get in ASAP (would be nice to get that small fix I ask for though)

But what I'm asking is that sanitycheck is taken out of the picture and we just automatically enable ASAN on the native port if CONFIG_TEST=y

aescolar commented:
> But what I'm asking is that sanitycheck is taken out of the picture and we just automatically enable ASAN on the native port if CONFIG_TEST=y

If it is failing at random now, I would not enable it by default in tests.

tyrantlizard77 reviewed (changes requested):
If there are random failures associated with this, we need to fix that.

So what I would like to see is:
1) Any random failures associated with ASAN resolved
2) libasan added to host dependency installation documentation
3) Through Kconfig enable ASAN if CONFIG_TEST=y and we're on a platform which supports ASAN

Special-casing sanitycheck like what has been proposed here has historically led to much wasted developer time and confusion when local builds / local sanitycheck runs / CI sanitycheck runs are not using the same options.

aescolar commented:
@andrewboie  About the failures see see #20122
(I also prefer the idea of keeping native_posix dependencies to a minimum by default)

vanwinkeljan commented:
@andrewboie Updated PR to address your 3 concerns

@aescolar 
> (I also prefer the idea of keeping native_posix dependencies to a minimum by default)

I do understand you like to keep the native_posix dependencies minimal but  as libasan is not an exotic requirement (is even part of GCC from 4.8 and onwards) I hope you don't mind I added an extra requirement.

vanwinkeljan commented:
Pushed update to sanitycheck script to disable ASAN when running sanitycheck with valgrind enabled as both ASAN and valgrind are not supported at the same time.

aescolar commented on subsys/debug/asan.c:
```diff
@@ -0,0 +1,37 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_SOC_POSIX)
```
Why are you switching on SOC_POSIX instead of the POSIX ARCH?

aescolar replied:
If you only hook into the asan library with __asan_default_options then it should be better to use `CONFIG_ARCH_POSIX`.
But if you add a native task, the part where you add the native task would be better be under a `#if defined(CONFIG_SOC_POSIX)`

aescolar commented on subsys/debug/Kconfig:
```diff
@@ -49,6 +49,7 @@ config TRACING
 config ASAN
 	bool "Build with address sanitizer"
 	depends on ARCH_POSIX
+	default y if TEST
```
So far you were speaking about enabling ASAN for hte native_posix board, but this enables it for all POSIX Arch boards. If anything this should be only on NATIVE_POSIX (& 64) not for all

aescolar replied:
> Do you see an issue to enable ASAN for all POSIX arch boards if they are running test?

Yep. Given that the nrf52_bsim is onyl relevant for BT developers, and has a very different aim than native_posix I would really prefer to not enable ASAN for it by default. So please do change this default to only apply to native_posix(64)

aescolar commented on boards/posix/native_posix/doc/index.rst:
```diff
@@ -40,6 +40,11 @@ but it has only been tested on Linux.
+.. note::
+   To run the test cases, either via sanitycheck script or manual via
+   west/cmake, you must have a version of libasan installed (in Ubuntu 18.04
+   install the libasan4 package).
```
Please add this dependency to the list of packages in the getting started guide. Too many people are not going to read this note.

SebastianBoe replied:
Most users will not use this so this should not be installed for all.

I'm sure we can find a way to give a reasonable error message when it's not installed.

vanwinkeljan replied:
@aescolar @SebastianBoe FYI, there is no need to update the `getting started guide` for ubuntu systems as the `gcc` packages already has a dependency on ASAN.

aescolar commented on subsys/debug/asan.c:
```diff
@@ -16,3 +21,14 @@ const char *__asan_default_options()
+
+static void lsan_do_leak_check(void)
+{
+	__lsan_do_leak_check();
```
Will this bite us back due to this interface not being stable?

aescolar replied:
Wasn't this necessary only for 64 bit targets?
(Not having looked into ASAN leak detection) Note that leak checks are not really useful anyhow for Zephyr on native_posix, as Zephyr's heap allocation is not using the Linux heap.
Leak detection can help you only with HW models or native_posix drivers leaks, but at this point (NATIVE_TASK ON_EXIT) there can still be some memory allocated by the board / HW models.
Meaning, would it maybe be easier to just disable ASAN leaks detection always?

vanwinkeljan replied:
> Will this bite us back due to this interface not being stable?

The interface has been there for a while and is both implemented by GCC and LLVM, so I assume it will not disappear quickly.

> Wasn't this necessary only for 64 bit targets?

Actually it is only need for 64-bit when building with GCC, but I rather keep the behavior the same for GCC and LLVM. Note that leak detection is not supported in 32-bit mode by ASAN.

>  Note that leak checks are not really useful anyhow for Zephyr on native_posix, as Zephyr's heap allocation is not using the Linux heap.

This is not 100% correct, a zephyr application can use malloc/free from libc and on a posix arch board this will result in the use of linux's malloc/free.

> Leak detection can help you only with HW models or native_posix drivers leaks, but at this point (NATIVE_TASK ON_EXIT) there can still be some memory allocated by the board / HW models.

This is true but we could add a suppression file.

> Meaning, would it maybe be easier to just disable ASAN leaks detection always?

I rather keep this functionality as zephyr apps could use malloc/free through libc.

aescolar reviewed (dismissed):
I guess I need to give up on the idea of not adding more dependencies to native_posix...
Anyhow a few issues below

aescolar commented on subsys/debug/CMakeLists.txt:
```diff
@@ -5,4 +5,9 @@ zephyr_sources_ifdef(
   openocd.c
   )
 
+zephyr_sources_ifdef(
```
Nit: This file is so posix arch dependent that I would only include if ARCH_POSIX

vanwinkeljan replied:
Actually `CONFIG_ASAN` already takes this in account through Kconfig `depends on`.
So I rather keep it a simple here; just in case the `depends on` changes in the Kconfig, else we need to update it in 2 places.

aescolar replied:
Fair enough

aescolar commented on subsys/debug/asan.c:
```diff
@@ -0,0 +1,37 @@
+		 * 64-bit boards if GCC is used.
+		 * https://github.com/zephyrproject-rtos/zephyr/issues/20122
+		 */
+		"leak_check_at_exit=0:"
```
If we are setting this one, why do we need to also call `__lsan_do_leak_check()`?

vanwinkeljan replied:
This flag prevents to run the leak checking `at exit` time in the global destructor which can lead to a deadlock. Note that leak detection bookkeeping logic is not disabled by the flag `leak_check_at_exit` and leak detection reporting can still be triggered via `__lsan_do_leak_check()`.

Further if `__lsan_do_leak_check()` is executed `at exit` leak detection is skipped and it could be argued that setting the default flag is not needed but this could lead to a deadlock if for some reason `__lsan_do_leak_check()` is not executed.

aescolar replied:
Ok.. so you want to keep doing the leak check but in a kinda bad place...

aescolar commented:
@andrewboie @nashif May I ask what does your approval cover? The sanitycheck changes alone? The way we are hooking into the ASAN library?

aescolar commented on subsys/debug/asan.c:
```diff
@@ -0,0 +1,37 @@
+
+static void lsan_do_leak_check(void)
+{
+	__lsan_do_leak_check();
```
Note: I'm still not too comfy with hooking into this asan interface always.

Defining __asan_default_options would seem harmless enough even if the interface changes, but calling this one would seem to be more risky, as it not only depends on the interface not being changed, but also depends on what [this function does](https://github.com/gcc-mirror/gcc/blob/master/libsanitizer/include/sanitizer/lsan_interface.h#L43).

Anyhow, please see my question above about it being needed at all.

vanwinkeljan replied:
I can see your concern but if this code is removed we lose the leak detection support from ASAN for native_posix_64 boards if they are build with GCC.

aescolar replied:
Yep.. but note my previous comment about its limited value

vanwinkeljan commented:
@aescolar I have the feeling that the pure ASAN functionality is accepted and the only remainder is the leak detection functionality (LSAN) and to be more precise the call to `__lsan_do_leak_check`.

Shall I move this part to a separate PR?

aescolar commented:
> @aescolar I have the feeling that the pure ASAN functionality is accepted and the only remainder is the leak detection functionality (LSAN) and to be more precise the call to __lsan_do_leak_check.

Yep, my only concern left is about the `__lsan_do_leak_check()` part.
There is also the setting CONFIG_ASAN by default only for native_posix (and not for all boards) which I understand you agree with and will fix.
And if you remove the __lsan_do_leak_check native task, then that new file ifdef which should be changed to CONFIG_ARCH_POSIX instead of SOC_POSIX, which I'm guessing you are ok with?

( there was the dependency by default to libasan, but I gave up about it)
So if you remove from this PR the "debug: native: Run LSAN leak check in ON_EXIT NATIVE_TASK" commit, and do those 2 other minor changes, I'm perfectly happy.

vanwinkeljan commented:
@aescolar addressed your concerns

vanwinkeljan commented:
Updated PR to add myself as code owner of `subsys/debug/asan.c`

mped-oticon reviewed (dismissed):
I'm all for these sanity checks, however Oticon is using CentOS and even with softwarecollections.org, a usable libasan is unavailable in a 32 bit (i686) version. See below.

This one sample point (of CentOS) will imply that other Linuxen/Mac is likely to experience dependency issues as well.

That renders this nice-to-have into something that will cause breakage.
Hence the default should be `CONFIG_ASAN=n`.

I am afraid the state of many 32-bit libraries/packages are neglected in many distros, so it would be wise to limit our dependence on them (as the default at least).

```
$ sudo yum search libasan
devtoolset-3-libasan-devel.x86_64 : The Address Sanitizer static library
devtoolset-4-libasan-devel.x86_64 : The Address Sanitizer static library
devtoolset-6-libasan-devel.x86_64 : The Address Sanitizer static library
devtoolset-7-libasan-devel.x86_64 : The Address Sanitizer static library
devtoolset-8-libasan-devel.x86_64 : The Address Sanitizer static library
libasan.i686 : The Address Sanitizer runtime library
libasan.x86_64 : The Address Sanitizer runtime library
libasan-static.i686 : The Address Sanitizer static library
libasan-static.x86_64 : The Address Sanitizer static library
libasan2.x86_64 : The Address Sanitizer runtime library from GCC 5
libasan3.x86_64 : The Address Sanitizer runtime library from GCC 6
libasan4.x86_64 : The Address Sanitizer runtime library from GCC 7
libasan5.x86_64 : The Address Sanitizer runtime library from GCC 8
```

aescolar commented:
@mped-oticon Thanks for checking. That is what I was fearing.
RedHat (CentOs) is a very normal distro in companies. So this is quite a deal breaker :(

vanwinkeljan commented:
We could only enable it by default for native_posix_64 and further set native_posix_64 to run by default in CI (and disable native_posix)

aescolar commented:
> We could only enable it by default for native_posix_64

That reduces the likehood of issues (at least it fixes this one in particular).. but still..

> further set native_posix_64 to run by default in CI (and disable native_posix)

That part nope, by far most of Zephyr targets are 32bit, and quite a few bugs are dependent on word-width. So the default should be the 32bit native_posix.

vanwinkeljan commented:
Then maybe just keep it "as is" and in case somebody wants to run sanitycheck without ASAN they can use the command `sanitycheck -x CONFIG_ASAN=n`?

aescolar commented:
> Then maybe just keep it "as is" and in case somebody wants to run sanitycheck without ASAN they can use the command sanitycheck -x CONFIG_ASAN=n?

.. if we merge it as is, then they would also need to override each time they build something that sets CONFIG_TEST.
The thing is that if native_posix by default does not work in somebodies environment, people will either assume it is broken and not use it, or file a bug report.

vanwinkeljan commented:
> .. if we merge it as is, then they would also need to override each time they build something that sets CONFIG_TEST.

Correct, but the only way around this is to set the ASAN config flag in the sanitycheck script (or CI run script), but there is an objection of @andrewboie against this solution as there will be a difference between local builds and sanitycheck builds.

So what is your suggestion to close this PR and let new bugs sneak in to the code base*?

(*) I'm still open to make it 64-bit only feature, but then I prefer to enable running CI again for native_posix_64 in that case else we will see issues popping up in the nightly runs (could go in a separate PR)

aescolar commented:
> So what is your suggestion to close this PR and let new bugs sneak in to the code base*?

I do not have any good way forward, but maybe the 64 bit only option is the best one.
@mped-oticon could you check in your environment if you can build with ASAN in native_posix_64?

nashif commented:
suggestion: disable by default, and make this an option for sanitycheck that is also disabled by default, we can enable that in sanitycheck on-demand, pretty much similar to valgrind. This will let us enable the feature in sanitycheck, how we do this in CI is a different topic IMO.

vanwinkeljan commented:
> suggestion: disable by default, and make this an option for sanitycheck that is also disabled by default, we can enable that in sanitycheck on-demand, pretty much similar to valgrind. This will let us enable the feature in sanitycheck

That is exactly what PR #20160 is proposing, but it has a NAK from @andrewboie 

> how we do this in CI is a different topic IMO.

I could revert this PR so that it enables it in the run ci script

aescolar commented:
> I could revert this PR so that it enables it in the run ci script

It sound to me like we are in an infinite loop :)

nashif commented:
> It sound to me like we are in an infinite loop :)

not really, other PR merged.

vanwinkeljan commented:
> not really, other PR merged.

Problem is that this PR is about enabling ASAN by default during CI (or nightly run) to find issues in the code base (see issue #19012) and not about being able to enable ASAN as such from the sanitycheck script.

So we are not yet their.

aescolar commented:
> So we are not yet their.

Well.. making it easier for people to run sanitycheck w ASAN is already an improvement. It is not all, but quite a lot.

vanwinkeljan commented:
Moved fixes for #20122 to a seperate PR #20320

vanwinkeljan committed:
ci: Enable ASAN during CI run

Updated run_ci script to enable address sanitizer (ASAN)

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

vanwinkeljan commented:
Reverted PR back to run_ci.sh solution to enable ASAN during CI run

vanwinkeljan commented:
Added DNM label as @andrewboie is against proposed solution, will remove label once @andrewboie gave his -1

tyrantlizard77 reviewed (changes requested):
as I've said earlier, I am against having the CI runs build with different options than local.

andrewboie commented:
> I'm all for these sanity checks, however Oticon is using CentOS and even with softwarecollections.org, a usable libasan is unavailable in a 32 bit (i686) version. See below.

> This one sample point (of CentOS) will imply that other Linuxen/Mac is likely to experience dependency issues as well.

@mped-oticon This is not a problem that is going to get any better over time.
Mac OS Catalina cannot run 32-bit binaries at all. The major Linux distros are also considering dropping 32-bit packages, Ubuntu was about to for their next release until an outcry forced them to postpone it.

If you are insisting that 32-bit native posix be the default and the canonical target for testing, you need to either reconsider your plans or come up with a way to deal with this. It's just going to get worse.

> We could only enable it by default for native_posix_64 and further set native_posix_64 to run by default in CI (and disable native_posix)

I prefer this approach.

> That part nope, by far most of Zephyr targets are 32bit, 
> and quite a few bugs are dependent on word-width.

Please provide some evidence for this assertion, I'm not taking your word for it. It's true that the 64-bit targets were introduced this year and there was work to be done to correct issues because this simply was never tested before, but now that integration is done I don't think there's evidence that word width bugs continue to dominate, such that testing on a 32-bit native port outweighs the **considerable** disadvantages of trying to get 32-bit binaries to run on modern OSes

tl;dr: 32-bit native is a dead end. You need to deal with it.

aescolar commented:
> @mped-oticon This is not a problem that is going to get any better over time.

No, it is not. That is why I prefer to ensure that defaults dependencies allow as many people as possible to use it.
Today, if circumstances permit, you can compile with ASAN locally and run with it. But if you are prevented for some reason, you can still run without it and benefit from everything else native_posix provides. Most issues found by ASAN are found by valgrind, and both of this tools focus on just memory access types of bugs. So is not like without ASAN you can't do anything else.

Note that today, the only dependency native_posix really has is the the 32bit libC, which will need to be supported by distros while any other 32 bit library is. So hopefully for a good while.

>> That part nope, by far most of Zephyr targets are 32bit,
and quite a few bugs are dependent on word-width.

> @aescolar Please provide some evidence for this assertion, I'm not taking your word for it. 

I think you misunderstood me. I'm not referring to current bugs in-tree which would prevent us from running w 64 bits, but about new bugs and bugs in application code which would affect 32bits but be hidden in 64bits targets.
For example: Changing word-with from 32 to 64 bits changes your structures layout in memory, adding more alignment holes, and causing all pointers (and pointer arrays) to double in size, so bugs in which you are accessing an structure at the wrong offset (say overflowing an array inside a struct) can be hidden or exposed depending on this.
So my point is, given that we are targeting primarily 32 bit µC, it is logical that we try to have a test  environment which reproduces their conditions relatively closely whenever easily doable. And not dropping native_posix(_32) by default in CI is certainly easy to do.

Note that native_posix_64 is still run, it is just not run for all tests by default.
I'm fine with running (again) also native_posix_64 by default for all tests. We did that for a while, and it was removed only to save CI time (which is really build time, as the run time is nill)

vanwinkeljan commented:
@andrewboie 

> And consider this scenario: some poor developer submits a PR which fails CI due to some ASAN issue flagged, and they are unable to reproduce it locally because their workstation doesn't have a working ASAN package! What exactly are they supposed to do then?

This poor developer could still use the CI docker image.

@aescolar 

> Most issues found by ASAN are found by valgrind, and both of this tools focus on just memory access types of bugs.

Dont think that this is the case, have a look at the list of issues fixed in #19012 and see which one would have been detected by valgrind (most of them are overruns in the stack). Btw I'm not taking here about the LSAN (leak sanitizer) functionality that has been merged in ASAN, on that one I agree it would have been detected by valgrind.

> Note that native_posix_64 is still run, it is just not run for all tests by default.
I'm fine with running (again) also native_posix_64 by default for all tests. We did that for a while, and it was removed only to save CI time (which is really build time, as the run time is nill)

Now I'm tempted to push a new change enabling it by default and see how much longer the CI run takes (now it was 46 minutes)

aescolar commented:
> have a look at the list of issues

That is because a few times now I have run sanitycheck with valgrind and either fix or report whatever I found, so whatever you find with ASAN would be what is left ;)

vanwinkeljan committed:
boards: native_posix: Enable default testing on native_posix_64

Enable testing by default on native_posix_64 as this target has ASAN
enabled by default when running test cases.

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

vanwinkeljan commented:
Pushed extra commit enabling test run on native_posix_64 to see run duration impact and added DNM label as this change should not be merged.

vanwinkeljan commented:
Test run with native_posix_64 enabled did result in a runtime of 43 minutes

Test run _without_ native_posix_64: duration 46 minutes, 3100 tests, 348 skipped
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/55143/summary/console

Test run _with_ native_posix_64: duration 43 minutes, 3423 tests, 367 skipped
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/55433/summary/console

Out of these result I would conclude that enabling native_posix_64 by default has no impact on CI test duration.

@aescolar @andrewboie @nashif @galak my proposal is to enable ASAN for 64-bit targets (native_posix_64) by default if `CONFIG_TEST` is set and include native_posix_64 in CI runs. The extra requirement of libasan should not be an issue as it is part of both GCC and LLVM and should be commonly available on any modern linux system (and even Mac OS X).

andrewboie commented:
> @aescolar @andrewboie @nashif @galak my proposal is to enable ASAN for 64-bit targets (native_posix_64) by default if CONFIG_TEST is set and include native_posix_64 in CI runs. The extra requirement of libasan should not be an issue as it is part of both GCC and LLVM and should be commonly available on any modern linux system (and even Mac OS X).

thumbs up from me

vanwinkeljan commented:
@aescolar @nashif @galak any further input on my proposal if no objection I will update the PR accordingly?

aescolar commented:
@vanwinkeljan Let me give a try in RedHat to the 64 bit ASAN

aescolar commented:
On a RedHat 7.5 using devtoolset-7 :( :
```
* [...lpi/Zephyr/west_zephyr/zephyr/build]$ cmake ../samples/hello_world/ -DBOARD=native_posix_64 -GNinja -DCONFIG_ASAN=y
Zephyr version: 2.1.0
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.6.5", minimum required is "3.4") 
-- Selected BOARD native_posix_64
-- Loading /scratch/alpi/Zephyr/west_zephyr/zephyr/boards/posix/native_posix/native_posix_64.dts as base
Devicetree configuration written to /scratch/alpi/Zephyr/west_zephyr/zephyr/build/zephyr/include/generated/generated_dts_board.conf
Parsing Kconfig tree in /scratch/alpi/Zephyr/west_zephyr/zephyr/Kconfig
Loaded configuration '/scratch/alpi/Zephyr/west_zephyr/zephyr/boards/posix/native_posix/native_posix_64_defconfig'
Merged configuration '/scratch/alpi/Zephyr/west_zephyr/zephyr/samples/hello_world/prj.conf'
Merged configuration '/scratch/alpi/Zephyr/west_zephyr/zephyr/build/zephyr/misc/generated/extra_kconfig_options.conf'
Configuration saved to '/scratch/alpi/Zephyr/west_zephyr/zephyr/build/zephyr/.config'
-- The C compiler identification is GNU 7.3.1
-- The CXX compiler identification is GNU 7.3.1
-- The ASM compiler identification is GNU
-- Found assembler: /opt/rh/devtoolset-7/root/usr/bin/gcc
-- Cache files will be written to: /home/alpi/.cache/zephyr
-- Configuring done
-- Generating done
-- Build files have been written to: /scratch/alpi/Zephyr/west_zephyr/zephyr/build
* [...lpi/Zephyr/west_zephyr/zephyr/build]$ ninja
[75/75] Linking C executable zephyr/zephyr.elf
* [...lpi/Zephyr/west_zephyr/zephyr/build]$ zephyr/zephyr.exe
==2850==ERROR: AddressSanitizer failed to allocate 0xdfff0001000 (15392894357504) bytes at address 2008fff7000 (errno: 12)
==2850==ReserveShadowMemoryRange failed while trying to map 0xdfff0001000 bytes. Perhaps you're using ulimit -v
Aborted
```

vanwinkeljan commented:
@aescolar could you run `ulimit -v` on the system, could be that the system is limiting virtual memory usage?

Some background info:
https://github.com/google/sanitizers/wiki/AddressSanitizer#ulimit--v

aescolar commented:
@vanwinkeljan It is, `23000000`.
But all these issues indicate that turning ASAN by default is going to create more trouble than good.
I think it is best to close this PR, and focus on other matters that will clearly improve things for everybody (there is a very long list of issues in need of attention) instead of keeping arguing about trade-offs in this one.

vanwinkeljan commented:
> But all these issues indicate that turning ASAN by default is going to create more trouble than good.

So you rather let sneak in bugs in the code because there is a not widely available linux distribution that imposes limits? I'm assuming that the limit is set by the Redhat distro and not by an admin. Note that on Centos 7 there is no limit is set, so it looks Redhat specific.

aescolar commented:
> So you rather let sneak in bugs in the code because there is a not widely available

Yes, I prefer to not set too many dependencies in the default configuration so people can keep using it.

vanwinkeljan commented:
@aescolar @andrewboie 
Ok looks like we can not come to an agreement, expect the point we do not want to complicate the developers life over finding bugs in zephyrs code base. 

One final question could both of you live with running a CI job on **nightly** basis with ASAN enabled by means of passing `--enable-asan --enable-lsan` to sanitycheck?

aescolar commented:
Agreement has not been possible, and there has been 3 months inactivity => closing. Please reopen if you disagree.
(regarding your question, I do not have a strong opinion about having special runs in CI, but others may, so I refrained from answering)
